src/Domain/Model/Task/TaskRepository.php

```diff
@@ -0,0 +1,14 @@
+<?php
+
+namespace Domain\Model\Task;
+
+interface TaskRepository
```

Comment:
Read and write repositories are usually separated from each other. The write repository lives in the domain layer, as it contains all the business rules. The read repository lives in the application layer, available for other services/views that need to display data.

What's the reason you chose for read/write separation?

The one thing you haven't touched upon is actual interaction with your domain model, because you haven't implemented an actual domain rule. Currently, you can toggle `completed` in any situation (which is probably what you want), but what if you would actually have guard this rule.

For example: let's say a task can only be marked completed if it's after a not before date. With such a rule, the domain cannot publicly expose a mutable `completed` property, because someone could mark a task as completed even before the not before date. 

Another example (which actually is a domain rule now I think about it) is the setting the title. In your new `TaskWrite` model, you only validate the `title` in the constructor. I can currently set the title to an empty string again after initializing the model:

```php
$task = new TaskWrite(1, 'Test', false);
$task->title = ''; // Valid
```

Last but not least: you shouldn't be able to create a completed task. If you create a task, it should automatically set `completed` to `false`.

Your model is currently called an anemic domain model: https://martinfowler.com/bliki/AnemicDomainModel.html. Have a look at this article and we can have a discussion about this later 🙂 .

Reply:
> What's the reason you chose for read/write separation?

Being honest, I was not very confident that organised my read/write model and repo's in the "correct" way from the start. There was not a real strong reason, other than the have a write model which properties you can change and a read modal, which properties you can't change, for reasons you mentioned above.

> The one thing you haven't touched upon is actual interaction with your domain model, because you haven't implemented an actual domain rule. 

That's true, and I think that's more to come later when we introduce more features of interacting with tasks.
The domain rule you mentioned above of not being able to create a completed task could be implemented already though.

> I can currently set the title to an empty string again after initializing the model. 

You're right! 🤯 

Finally:
Let's have a discussion about the article tomorrow, because being honest I don't completely understand the article.
And I want to understand!